run Doxygen and doc coverage test on static Travis config

3nids opened the pull request.

3nids committed:
run Doxygen and doc coverage test on static Travis config

m-kuhn commented on cmake/CreateQgsVersion.cmake:
```diff
@@ -0,0 +1,47 @@
+MACRO(CREATE_QGSVERSION)
```
Can you add a header with description of what this does?

m-kuhn commented on doc/CMakeLists.txt:
```diff
@@ -46,65 +46,65 @@ IF(WITH_APIDOC)
-      ${CMAKE_SOURCE_DIR}/src/analysis/raster
-      ${CMAKE_SOURCE_DIR}/src/analysis/vector
-      ${CMAKE_SOURCE_DIR}/src/plugins
+      ${CMAKE_CURRENT_SOURCE_DIR}/../doc
```
Why did you change this to relative paths? I recently changed all the include paths to absolute (relative to `CMAKE_SOURCE_DIR`) and this commit here seems to go in the opposite direction.

3nids replied:
The issue is that this file is included from `.ci/travis/code_layout/CMakeLists.txt` hence the cmake top source directory is not the same.
In general, I would prefer to use the relative path for this reason, it allows more flexibilty.
What do you think?

m-kuhn replied:
It's kind of ok here.

But you soon go 3 levels up and then it requires a lot of brainpower to determine, where the target is (`../../core/`: was it the core directory in src/, python/, tests/? etc.).
Another drawback is, that when adding a new folder (e.g. the subfolder `src/core/expressions`) you can't just copy-paste this between different CMakeLists but you have to manually check for each and every single CMakeLists file the relative path.

Personally, I find these relative paths in CMakeLists a PITA to maintain and would prefer to switch it all to "relative to source root". And move this CMakeLists file here also to the root.

So I wonder if there is a possibility to move the logic from this CMakeLists file here to the main one or move all code related to this one to this subdirectory.

3nids replied:
That means having a config which does not build any QGIS part at all.
That means a huge part of the main CMakeLists would be wrap in something like `IF (WITH_CORE)`
Do you think it's better? 
I would say yes, since we have a single starting point for any build.

m-kuhn replied:
Yes, that's probably what it means. I agree that a single entry point sounds better.

3nids committed:
fix test import

3nids committed:
run code_layout build from top CMakeLists

instead of adding an extra CMakeLists in .ci/travis/code_layout to build API doc, astyle and run tests (indentation, spelling, sip, doc coverage), the top CMakeLists has been adapted to allow not building core libraries and possibly just the static code layout
* astyle has been moved from /src/astyle to /lib/astyle (I would propose to move all external libraries, and possibly add git submodules)

3nids commented:
I think this is ready to be merged.

* I added a `WITH_CORE` option which basically allows not to build QGIS. If you use `WITH_ASTYLE`, `WITH_API_DOC` and `ENABLE_TESTS` this will allow the static tests to be run on source code.
* astyle has been moved from `/src/astyle` to `/lib/astyle` (side remark: wouldn't it be better to have all external libs there, with maybe git submodules?)
* `.ci/travis/code_layout/CMakeLists.txt` has been removed since all its functionality were brought into top CMakeLists.txt
* for the moment, sip and doc coverage tests are still run in linux build until checking that new classes have their sip counterparts (ongoing [discussion](https://lists.osgeo.org/pipermail/qgis-developer/2017-June/048740.html))

@jef-n @m-kuhn any objection?

m-kuhn commented on tests/code_layout/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+  ADD_PYTHON_TEST(PyQgsDocCoverage ${CMAKE_SOURCE_DIR}/tests/code_layout/test_qgsdoccoverage.py)
+  IF (WITH_CORE)
+    ADD_PYTHON_TEST(PyQgsSipCoverage ${CMAKE_SOURCE_DIR}/tests/code_layout/test_qgssipcoverage.py)
+   ENDIF(WITH_CORE)
```
Bad indentation

m-kuhn commented on CMakeLists.txt:
```diff
@@ -34,298 +34,305 @@ SET(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake ${CMAKE_MODULE_PATH})
-
-# Custom widgets
-SET (WITH_CUSTOM_WIDGETS FALSE CACHE BOOL "Determines whether QGIS custom widgets for Qt Designer should be built")
+SET (WITH_CORE TRUE CACHE BOOL "Determines whether QGIS core should be built.")
```
Can you mark that as advanced? (`MARK_AS_ADVANCED(WITH_CORE)`)

m-kuhn commented:
Looks good.
I like the move to place astyle to a separate place. I wonder if a directory 3rdparty or external would be more verbose than lib?

3nids committed:
fix external lib path
